File: Wpf2DGrafik/ShapeFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using wpf = System.Windows;

namespace Wpf2DGrafik
{
    /// <summary>
    /// Basisklasse für Algorithmen zum Erzeugen von 2d- Grafikprimitiven mit der Maus
    /// </summary>
    public abstract class ShapeFactory
    {
        public static string MakeCtrlId()
        {
            return "N" + Guid.NewGuid().ToString().Replace('{', '_').Replace('}', '_').Replace('-', '_');
        }

        protected int _StepCount;
        protected int BuildStepCount
        {
            get
            {
                return _StepCount;
            }
        }

        int _Step = 0;
        public int BuildStep
        {
            get
            {
                return _Step;
            }
        }

        // Liste der Stützstellen, auf denen die Figur erzeugt wird (z.B. Start & Endpunkt einer Linie)
        protected List<wpf.Point> bases = new List<wpf.Point>();

        // Mit der Methode werden alle Stützpunkte der Figur erfasst. Nach erfassen des letzten 
        // Stützpunktes wird die Figur erzeugt und der Zeichnung hinzugefügt
        public bool CollectBaseAndCreateShape(wpf.Point pos, Zeichnung drawing)
        {
            if (_Step < _StepCount - 1)
            {
                _Step += 1;
                bases.Add(pos);
                return false;
            }
            else
            {
                bases.Add(pos);
                drawing.Elemente.Add(createShape());

                // Für nächste erfassung von Stützpunkten bereitmachen
                _Step = 0;
                bases.Clear();
                return true;
            }
        }


        // In einer abgeleiteten Klasse wird in dieser virtuellen Methode über den Stütztstellen 
        // die eigentliche Figur erzeugt
        protected abstract wpf.Shapes.Shape createShape();
    }
}

File: WpfGrafik/ShapeFactory.vb
Imports wpf = System.Windows
' Basisklasse für Algorithmen zum Erzeugen von 2d- Grafikprimitiven mit der Maus
Public MustInherit Class ShapeFactory

    Protected _StepCount As Integer
    Protected ReadOnly Property BuildStepCount As Integer
        Get
            Return _StepCount
        End Get
    End Property

    Dim _Step As Integer = 0
    Public ReadOnly Property BuildStep As Integer
        Get
            Return _Step
        End Get
    End Property

    ' Liste der Stützstellen, auf denen die Figur erzeugt wird (z.B. Start & Endpunkt einer Linie)
    Protected bases As New List(Of wpf.Point)

    ' Mit der Methode werden alle Stützpunkte der Figur erfasst. Nach erfassen des letzten 
    ' Stützpunktes wird die Figur erzeugt und der Zeichnung hinzugefügt
    Public Function CollectBaseAndCreateShape(ByVal pos As wpf.Point, ByVal drawing As Zeichnung) As Boolean
        If _Step < _StepCount - 1 Then
            _Step += 1
            bases.Add(pos)
            Return False
        Else
            bases.Add(pos)
            drawing.Elemente.Add(createShape())

            ' Für nächste erfassung von Stützpunkten bereitmachen
            _Step = 0
            bases.Clear()
            Return True
        End If
    End Function


    ' In einer abgeleiteten Klasse wird in dieser virtuellen Methode über den Stütztstellen 
    ' die eigentliche Figur erzeugt
    Protected MustOverride Function createShape() As wpf.Shapes.Shape

End Class
